### PR TITLE
[FCE-207] Fix error on muted track renegotiation

### PR DESCRIPTION
**Pull request by kamil-stasiak**
## Description

This PR fixes an error that occurs when a muted track is renegotiated.

## Motivation and Context

Steps to reproduce:
- Join the room with Peer 1.
- Join the room with Peer 2.
- Peer 2 mutes the microphone.
- Peer 2 turns off the camera.
- The error appears for Peer 1.

## Types of changes

Breaking change (fix or feature that would cause existing functionality to not work as expected)
- type `AudioOrVideoType` from `react-client` has been replaced by `TrackKind` from `ts-client`

### CODE REVIEW

File: packages/react-client/src/DeviceManager.ts [hunk 1]
```diff
@@ -164,7 +164,7 @@ const handleNotAllowedError = async (constraints: MediaStreamConstraints): Promi
   return await getMedia({ video: false, audio: false }, { video: PERMISSION_DENIED, audio: PERMISSION_DENIED });
 };
 
-const getError = (result: GetMedia, type: AudioOrVideoType): DeviceError | null => {
+const getError = (result: GetMedia, type: TrackKind): DeviceError | null => {
```
Comment (karkakol) [inline packages/react-client/src/DeviceManager.ts:167]: Imo `TrackType` sounds better than `TrackKind`

File: packages/ts-client/src/webrtc/transceivers.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+      return acc
+    }, {} as Record<string, string>);
+
+
+export const getMuted = <EndpointMetadata, TrackMetadata>(
+  midToTrackId: Map<string, string> = new Map(),
+  localEndpoint: EndpointWithTrackContext<EndpointMetadata, TrackMetadata>,
+): Record<string, string> => {
+  return [...midToTrackId.entries()]
```
Comment (karkakol) [inline packages/ts-client/src/webrtc/transceivers.ts:42]: midToTrackId.entries() would not be enough?

File: packages/ts-client/src/webrtc/transceivers.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+): Record<string, string> | null =>
+  connection.getTransceivers()
+    .filter((transceiver) => transceiver.sender.track?.id && transceiver.mid)
+    .reduce((acc, transceiver) => {
+      const localTrackId = transceiver.sender.track!.id;
+      const mid = transceiver!.mid!;
+
+      const trackContext = Array.from(localTrackIdToTrack.values()).find(
+        (trackContext) => trackContext?.track?.id === localTrackId,
+      )!;
```
Comment (Rados13) [inline packages/ts-client/src/webrtc/transceivers.ts:30]: Could we move it to a function?

File: packages/ts-client/src/webrtc/transceivers.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  localEndpoint: EndpointWithTrackContext<EndpointMetadata, TrackMetadata>,
+): Record<string, string> => {
+  return [...midToTrackId.entries()]
+    .filter(([_mid, trackId]) => localEndpoint.tracks.get(trackId))
+    .reduce((acc, [mid, trackId]) => {
+      acc[mid] = trackId;
+
+      return acc
+    }, {} as Record<string, string>);
+}
```
Comment (Rados13) [inline packages/ts-client/src/webrtc/transceivers.ts:49]: If I understand correctly this function should return muted tracks, where do we filter out tracks that are not muted here?

File: packages/react-client/src/DeviceManager.ts (hunk 1 above)
Comment (czerwiukk) [reply packages/react-client/src/DeviceManager.ts:167]: maybe update the variable name as well to `kind: TrackKind`

Comment (karkakol) [reply packages/react-client/src/DeviceManager.ts:167]: Agreed with @czerwiukk. lets rename function parameter name to `kind`.

File: packages/ts-client/src/webrtc/webRTCEndpoint.ts [hunk 5]
```diff
@@ -1607,13 +1588,16 @@ export class WebRTCEndpoint<
...
+
       if (this.checkIfTrackBelongToEndpoint(trackId, this.localEndpoint))
         return;
+      if (!isTrackKind(event.track.kind)) throw new Error("Track has no kind")
 
       const trackContext = this.trackIdToTrack.get(trackId)!;
 
       trackContext.stream = stream;
       trackContext.track = event.track;
+      trackContext.trackKind = event.track.kind;
```
Comment (czerwiukk) [inline packages/ts-client/src/webrtc/webRTCEndpoint.ts:1600]: i may have missed something, but why are we copying the `event.track.kind` information to another field while we can access it through `trackCotnext.track.kind`?

File: packages/ts-client/src/webrtc/transceivers.ts [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+): Record<string, string> | null => {
+  if (!connection) return null;
+
+  const active = getActive(connection, localTrackIdToTrack)
+  const muted = getMuted(midToTrackId, localEndpoint)
+
+  return { ...active, ...muted };
+};
+
+export const getActive = <EndpointMetadata, TrackMetadata>(
```
Comment (czerwiukk) [inline packages/ts-client/src/webrtc/transceivers.ts:18]: what is active? isn't the name lacking something?

Comment (czerwiukk) [reply packages/ts-client/src/webrtc/transceivers.ts:18]: looking at the code and comments, i'd go for getActiveTracks/mutedTracks

File: packages/ts-client/src/webrtc/transceivers.ts (hunk 2 above)
Comment (kamil-stasiak) [reply packages/ts-client/src/webrtc/transceivers.ts:42]: I need it to map the result of `entries()`, which is an `IterableIterator`, to a simple array.

File: packages/ts-client/src/webrtc/webRTCEndpoint.ts (hunk 5 above)
Comment (kamil-stasiak) [reply packages/ts-client/src/webrtc/webRTCEndpoint.ts:1600]: The `track` field is optional; it's possible to mute a track (replace the track with null) and then unmute it in the future. We need the `trackKind` to assess bandwidth even though the track is muted

File: packages/ts-client/src/webrtc/transceivers.ts (hunk 4 above)
Comment (kamil-stasiak) [reply packages/ts-client/src/webrtc/transceivers.ts]: I've changed the names of these functions in this file, and I've added some comments

File: packages/ts-client/src/webrtc/transceivers.ts [hunk 7]
```diff
@@ -1,49 +1,59 @@
...
+): MidToTrackId | null => {
   if (!connection) return null;
 
-  const active = getActive(connection, localTrackIdToTrack)
-  const muted = getMuted(midToTrackId, localEndpoint)
+  // - negotiated unmuted tracks: tracks added in previous negotiation, data is being transmitted
+  // - not yet negotiated tracks: tracks added in this negotiation, data will be transmitted after successful negotiation
+  const transceiver = getTransceiverMapping(connection, localTrackIdToTrack)
+
+  // - negotiated unmuted tracks: tracks added in previous negotiation, data is being transmitted
```
Comment (Rados13) [inline packages/ts-client/src/webrtc/transceivers.ts:20]: Why do we have the same comment two times?

Comment (kamil-stasiak) [reply packages/ts-client/src/webrtc/transceivers.ts:20]: Because it's an intersection of these two functions

**Review by Rados13 — APPROVED** [review]

**Pull request merged**